README.md
`respfun`
================

  - [Installation](#installation)
  - [Functions](#functions)
      - [`split_rate`](#split_rate)
      - [`eff_vol`](#eff_vol)
      - [`spec_density`](#spec_density)
      - [`wm_to_vol`](#wm_to_vol)
      - [`scale_rate`](#scale_rate)
      - [`q_ten`](#q_ten)
      - [`flush_time`](#flush_time)
  - [Future functionality](#future-functionality)
  - [Bug reports](#bug-reports)

<!-- README.md is generated from README.Rmd. Please edit that file -->

[![Travis build
status](https://travis-ci.org/nicholascarey/respfun.svg?branch=master)](https://travis-ci.org/nicholascarey/respfun)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/nicholascarey/respfun?branch=master&svg=true)](https://ci.appveyor.com/project/nicholascarey/respfun)
[![Coverage
status](https://codecov.io/gh/nicholascarey/respfun/branch/master/graph/badge.svg)](https://codecov.io/github/nicholascarey/respfun?branch=master)
[![DOI](https://zenodo.org/badge/174339770.svg)](https://zenodo.org/badge/latestdoi/174339770)

The `respfun` package is a collection of functions for use with
respirometry data and experiments. I will add to it periodically
([suggestions
welcome](https://github.com/nicholascarey/respfun/issues)). It is not
intended to be a fully featured package, more a collection of handy
functions.

If you use it a citation using this [Zenodo
DOI](https://zenodo.org/record/3668902) would be much appreciated.

For full analyses of your respirometry data check out another package I
co-developed: [`respR`](https://github.com/januarharianto/respR).

### Installation

`respfun` can be installed using the `devtools` package:

``` r
install.packages("devtools")
devtools::install_github("nicholascarey/respfun")
```

### Functions

Currently there are seven functions:

#### `split_rate`

This function divides a metabolic rate between a group of individuals
based on their masses and a metabolic scaling exponent. It was used in
this recent publication:

*Benjamin P. Burford, Nicholas Carey, William F. Gilly, Jeremy A.
Goldbogen. 2019. Grouping reduces the metabolic demand of a social
squid. **Marine Ecology Progress Series**. 612: 141–150
<https://doi.org/10.3354/meps12880>*

Rates can be entered manually, and it also integrates with the
[`respR`](https://github.com/januarharianto/respR) package: objects
saved from the `respR::convert_rate` function can be entered (or `%>%`
piped), and the rate and units will be automatically extracted.

``` r
## Simple example

split_rate(tR = 500,                  # total metabolic rate of group
           masses = c(2, 3, 4, 5, 6), # body masses
           b = 0.75,                  # metabolic scaling exponent
           units = "mg/h")            # units (optional)
```

    #> 
    #> # split_rate # -------------------------
    #> Rate Division Complete: 
    #> 
    #> Total Group Rate(s) ($tR, entered):           
    #> [1] 500
    #> 
    #> Metabolic Scaling Exponent ($b, entered):  
    #> [1] 0.75
    #> 
    #> Masses ($masses, entered): 
    #> [1] 2 3 4 5 6
    #> 
    #> Intercept(s) ($a, calculated) :               
    #> [1] 35.79844
    #> 
    #> Individual rates ($indiv.rates, calculated): 
    #> [[1]]
    #> [1]  60.20557  81.60281 101.25329 119.69931 137.23902
    #> 
    #> Rate units: mg/h

``` r
## respR example

library(respR)                                           # Load respR

urchins.rd %>%                                           # NOT a group respirometry experiment,
                                                          # - Just using it as an example,
  inspect(1, 15) %>%                                     # inspect
  calc_rate(from = 4, to = 29, by = "time") %>%          # calculate rate
  print() %>%
  convert_rate(o2.unit = "mgl-1", time.unit = "m",       # convert
               output.unit = "mg/h", volume = 1.09) %>%
  split_rate(c(2, 3, 4, 5, 6, 7, 8),                     # divide
             0.75) %>%
  print()
```

    #> 
    #> # calc_rate # -------------------
    #> Rate(s):
    #> [1] -0.02177588
    #> 
    #> # convert_rate # ------------------------
    #> Rank/position 1 result shown. To see all results use summary().
    #> Input:
    #> [1] -0.02177588
    #> [1] "mg/L" "min" 
    #> Converted:
    #> [1] -1.424143
    #> [1] "mg/hour"
    #> 
    #> # split_rate # -------------------------
    #> Rate Division Complete: 
    #> --- respR::convert_rate object detected ---
    #> 
    #> Total Group Rate(s) ($tR, entered via convert_rate input):
    #> [1] -1.424143
    #> 
    #> Metabolic Scaling Exponent ($b, entered):  
    #> [1] 0.75
    #> 
    #> Masses ($masses, entered): 
    #> [1] 2 3 4 5 6 7 8
    #> 
    #> Intercept(s) ($a, calculated) :               
    #> [1] -0.06184548
    #> 
    #> Individual rates ($indiv.rates, calculated): 
    #> [[1]]
    #> [1] -0.1040113 -0.1409772 -0.1749254 -0.2067928 -0.2370945 -0.2661531 -0.2941883
    #> 
    #> Rate units: mg/hour

See
[`split_rate`](https://nicholascarey.github.io/respfun/reference/split_rate.html)
for more.

#### `eff_vol`

This calculates the ‘effective volume’ of a respirometer, that is the
volume *minus* the volume of the specimen. This allows you to input the
correct water volume into respirometry analyses in order to calculate
the correct amount of oxygen used, for example in the
[`respR::convert_rate`](https://januarharianto.github.io/respR/reference/convert_rate.html)
function.

The respirometer volume (`resp_vol`) can be corrected to get the
effective volume in a number of ways:

  - A specimen volume (`spec_vol`) can be entered directly (if for
    instance you have calculated it geometrically or measured the
    displacement volume separately), in which case the effective volume
    is a simple subtraction.

  - Alternatively, you can enter the specimen mass and density (see
    `spec_density` function below), in which case the specimen volume is
    calculated and the correction performed using that. If your specimen
    is neutrally buoyant you could enter the density of the water here
    (seawater density is usually around 1026 kg/m^3, but see next option
    for precise calculation).

  - Lastly, you can enter the specimen mass and make the assumption the
    specimen is the same density as the water, in which case temperature
    and salinity are required to calculate the water density and perform
    the correction (strictly speaking atmospheric pressure is also
    required; in reality it has a negligible effect within normal
    ranges, although the default value can be changed if desired). This
    is common in fish respirometry when the specimen is known to be
    neutrally buoyant, or nearly so.

See
[`eff_vol`](https://nicholascarey.github.io/respfun/reference/eff_vol.html)
for more.

#### `spec_density`

This function calculates a specimen density from a total wet mass and a
bouyant mass (see
[here](https://www.researchgate.net/publication/266911357_Buoyant_weight_technique_Application_to_freshwater_bivalves)).
This is common in studies of shell bearing species such as molluscs and
echinoderms where it is a non-lethal way of examining the ratio of shell
to tissue mass. It can also be used to determine the animal volume for
correcting respirometry water volumes (see above). The water temperature
and salinity at which buoyant mass was determined are required. See
[`spec_density`](https://nicholascarey.github.io/respfun/reference/spec_density.html)
for more.

#### `wm_to_vol`

This is a simple function to convert a water mass to a water volume.
Sometimes it is easier to weigh a respirometer to determine the water
mass than try to measure the volume, for example across large size
ranges of specimen and respirometer size (e.g. [Carey et
al. 2016](https://www.dropbox.com/s/d4zp3vm6xakzkts/Carey%20et%20al%20JEB%202016.pdf?dl=0)),
where systematic error is a concern. The water temperature and salinity
are required. See
[`wm_to_vol`](https://nicholascarey.github.io/respfun/reference/wm_to_vol.html)
for more.

#### `scale_rate`

This function scales a physiological rate to a different body mass using
a scaling exponent. Works with both absolute (i.e. whole animal) and
mass-specific rates. This function integrates with the
[`respR`](https://github.com/januarharianto/respR) package: objects
saved from the `respR::convert_rate` function can be entered (or `%>%`
piped), and the rate will be automatically extracted. See
[`scale_rate`](https://nicholascarey.github.io/respfun/reference/scale_rate.html)
for more.

#### `q_ten`

Calculates any of the five parameters in the Q10 temperature
relationship for physiological or chemical processes. Q10 describes the
ratio by which a physiological or chemical rate changes with a 10°C
increase in temperature. Essentially the same as the `Q10` function in
the
[`respirometry`](https://cran.r-project.org/web/packages/respirometry/index.html)
package by Matthew Birk, although that has additional functionality for
determining the best fit Q10 for a range of rates/temperatures. See
[`q_ten`](https://nicholascarey.github.io/respfun/reference/q_ten.html)
for more.

#### `flush_time`

Calculates the time taken to flush a respirometer with new water. It
requires the volume of the respirometer (in L) and the flow rate (in
L/s) of the input of new water. See
[`flush_time`](https://nicholascarey.github.io/respfun/reference/flush_time.html)
for more.

### Future functionality

In due course I’ll add a few more functions I find useful in working
with respirometry and metabolic rate data. Suggestions for additional
functions are welcome via [email](mailto:nicholascarey@gmail.com), or by
[opening an issue](https://github.com/nicholascarey/respfun/issues).

### Bug reports

If you find any bugs, weird behaviour, or other issues please [let me
know](mailto:nicholascarey@gmail.com) or [open an
issue](https://github.com/nicholascarey/respfun/issues).


NEWS.md
# respfun 0.4.4
Very minor update

- A startup message is now shown on package load

# respfun 0.4.3
Minor update

- `split_rate` now accepts multiple input rates, including from `convert_rate` objects. These should obviously all be rates from the *same group*. Minor change to output object as a result: `$indiv.rates` is now a `list` object with each element a vector of individual rates for teh associated group rate in `tR`. Extract via ``$indiv.rates[[1]]` etc.

# respfun 0.4.2
Another minor fix to `scale_rate`. 

- The `scale_rate` function now properly extracts all rates from `respR::convert_rate` objects, not just the first one. 

# respfun 0.4.1

This patch fixes a major issue in `eff_vol`. All previous versions of this function returned incorrect results at anything other than neutral buoyancy (multiplication instead of division in the mass & density to volume equation, doh!).

- MAJOR FIX: fix for incorrect results returned by `eff_vol`. If you have used this function previously, please review your code and results. 
- FIX:`spec_density` : warning when wet_mass and buoy_mass are equal (leads to infinite density calculation)



R/split_rate.R
#' @title Divide a metabolic rate amongst a group
#'
#' @description Divide a metabolic rate amongst a group of individuals.
#'
#' @details Divides a group metabolic rate amongst individuals in the group,
#'   given body masses of each and a scaling exponent.
#'
#'   Take care to enter the correct scaling exponent (`b`). This should be the
#'   scaling exponent of absolute metabolic rates, not mass-specific, and is
#'   usually (with exceptions) a positive value between 0.66 and 1. If your `b`
#'   value is less than this, especially if it is less than 0.33, and
#'   *especially* if it is negative, then it is likely a *mass-specific* scaling
#'   exponent. In which case, the correct scaling exponent is the positive
#'   difference of this value from 1. For example, an absolute scaling exponent
#'   of 0.75 has a mass-specific scaling exponent of -0.25.
#'
#'   If, for whatever reason, you want to do a simple per-capita division of the
#'   rate regardless of the body masses, enter `b = 0`. Or just divide it by the
#'   number of individuals.
#'
#'   Multiple rates can be entered, either as a vector or as part of a
#'   `respR::convert_rate` input for `tR`, but these should be separate
#'   measurements of rate of the same group. The output will contain separate
#'   individual rate vectors for each group rate, and the intercept (`a`) for
#'   each.
#'
#'   Units can be entered, e.g. `units = "mg/h"`, if the user wants to save
#'   these in the output for reference, however they do not affect calculations
#'   in any way. If a `respR::convert_rate` object is used for the rate (see
#'   below), the units are extracted from this and any other `units` input
#'   ignored.
#'
#' @section Sign of the rate: NOTE: both negative and positive rates can be
#'   entered. In respirometry experiments, rates are typically reported as
#'   positive values. These can be entered as is. In the case of
#'   `respR::convert_rate` objects, extracted rates will typically be
#'   *negative*, and these are left unchanged in the `split_rate` function. In
#'   `respR`, to be mathematically consistent (since they represent oxygen
#'   depletion), respiration rates are represented by negative slopes, and
#'   therefore rates returned as negative. You can enter the rate as the usually
#'   reported postive value, or as a negative: the function will work with
#'   either. Returned rates and intercept `a` will be identical in value, except
#'   for the sign. In effect, if you enter a negative rate you can simply ignore
#'   the signs for `a` and individual rates in the output.
#'
#' @section `respR` integration: For total rate (`tR`) the function accepts
#'   objects saved from the \code{respR}
#'   (\url{https://github.com/januarharianto/respR}) `convert_rate` function. In
#'   this case, the rate(s) and units are automatically extracted. However, if
#'   it contains a mass-specific rate (i.e. the rate has been adjusted in
#'   `respR::convert_rate` to a specific mass), **no conversion is done** and a
#'   warning is returned. Only absolute, that is non-mass specific, respiration
#'   rates should be divided in this way.
#'
#' @section Output: Output is a `list` object containing 7 elements:
#'
#'   `$a` = `a`, the intercept(s) in the mass~rate power equation. Determined by
#'   the function.
#'
#'   `$b` = `b`, the exponent in the mass~rate power equation. User entered.
#'
#'   `$tR` = total group rate. User entered.
#'
#'   `$masses` = specimen masses. User entered.
#'
#'   `$units` = units of the rate. User entered or extracted from `convert_rate`
#'   object. For information only, does not affect any calculations.
#'
#'   `$indiv.rates` = Primary output of interest, the group rate divided between
#'   individuals. This is a `list` object with each element a vector of
#'   individual rates associated with the `tR` at the same position. The sum of
#'   each vector should equal the associated `tR`. Extract via
#'   `$indiv.rates[[1]]` etc.
#'
#'   `$input` = origin of `tR` and `units`. Either `manual` entry or
#'   `convert_rate` object
#'
#' @usage split_rate(tR = NULL, masses, b = 0.75, units = NULL)
#'
#' @param tR numeric. Total group metabolic rate. Single value or vector of
#'   multiple measurements of rate of same group.
#' @param masses numeric. A vector of body masses of all individuals in group.
#' @param b numeric. Metabolic scaling exponent.
#' @param units string. Units of the rate. Extracted from `convert_rate` object
#'   or can be entered by the user. For information only, does not affect any
#'   calculations.
#'
#' @examples
#' # Single group rate
#' split_rate(tR = 500, masses = c(2, 3, 4, 5, 6), b = 0.75)
#'
#' # Multiple group rate
#' split_rate(tR = c(500,550,600), masses = c(2, 3, 4, 5, 6), b = 0.75)
#'
#' @author Nicholas Carey - \email{nicholascarey@gmail.com}
#'
#' @export

split_rate <- function(tR = NULL, masses, b = 0.75, units = NULL) {

  ## entry checks
  if (is.null(tR)) {
    stop("Enter tR - Total Group Rate")
  }
  if (is.null(b)) {
    stop("Enter value for b - Metabolic Scaling Exponent")
  }

  ## check b is within fairly normal range - but don't stop function
  ## just a warning in case user enters MASS SPECIFIC scaling exponent
  if ((b < 0.66 && b > 0.33) || b > 1) {
    warning("Value of b is outside the typical range for a metabolic scaling exponent of 0.66 to 1.
  Check this value is correct, and not the MASS-SPECIFIC scaling exponent.
  Result has been returned regardless.")
  }

  if (b <= 0.33 && b >= 0) {
    warning("Value of b is less/equal to 0.33, which is WELL outside the typical range of 0.66 to 1.
  Check this value is correct, and not the MASS-SPECIFIC scaling exponent.
  Result has been returned regardless.")
  }

  if (b < 0) {
    warning("Value of b is NEGATIVE. Typical b values are positive, between 0.66 and 1.
  Check this value is not the MASS-SPECIFIC scaling exponent.
  Result has been returned regardless.")
  }


# respR convert_rate object input -----------------------------------------

  if(class(tR) == "convert_rate"){
    input <- "convert_rate"


    if(!is.null(units))
      message("'units' input ignored. Rate units extracted from convert_rate object.")

    rate <- tR$output
    units <- tR$output.unit
    pattern <- c("*/ug", "*/mg", "*/g", "*/kg")

    if(any(sapply(pattern, function(x) grepl(x, units))))
      stop("Mass-specific units detected in convert_rate object.
  Cannot split rate by mass if rate is already corrected for mass!")
  }


# Numeric rate input ------------------------------------------------------

  if(is.numeric(tR))
    rate <- tR

  if(is.numeric(tR))
    input <- "manual"

  if(is.null(units))
    units <- "Undefined"

  ## calculate a
  a <- rate / (sum(masses^b))

  ## return scaled rates
  ## lapply in case multiple rates
  indiv_rates <- lapply(a, function(x) x * (masses^b))

  ## assemble output
  output <- list(
    a = numeric(0), b = numeric(0), tR = numeric(0),
    masses = vector(), units = character(), indiv.rates = vector(),
    input = character()
  )

  output[[1]] <- a
  output[[2]] <- b
  output[[3]] <- rate
  output[[4]] <- masses
  output[[5]] <- units
  output[[6]] <- indiv_rates
  output[[7]] <- input

  ## Assign class
  class(output) <- "split_rate"

  return(output)
}


